src/vanilla.ts

```diff
@@ -15,17 +15,15 @@ type Op =
   | [op: 'reject', path: Path, error: unknown]
 type Listener = (op: Op, nextVersion: number) => void
 
-type AnyFunction = (...args: any[]) => any
-
-type Snapshot<T> = T extends AnyFunction
-  ? T
-  : T extends AsRef
+type Snapshot<T> = T extends AsRef
   ? T
-  : T extends Promise<any>
+  : T extends Promise<unknown>
   ? Awaited<T>
-  : {
-      readonly [K in keyof T]: Snapshot<T[K]>
-    }
+  : T extends unknown[]
+  ? readonly Snapshot<T[number]>[]
+  : T extends { [key: string]: unknown }
```

Comment:
Fwiw classes don't extend `{ [key: string]: unknown }`, so this will stop `Snapshot` from recursing into classes.

Which is a good thing for types like `Date`, but doesn't technically match `snapshot`s runtime behavior, which will snapshot anything that passes `canProxy`.

Currently `canProxy` just special cases `Date` / `Number` / etc., and I think the mapped type will probably have to do that as well (?) if it wants to exactly line up the type system behavior to the runtime behavior.

(That is typically how I've seen this done, like a `type Primitive = string | number | Date | ...` and then `T extends Primitive ? T : ...recurse...`, but I was actually hoping this would work b/c it's much smaller/more succinct)

Reply:
Hmm. Yes, you're right. Thanks for pointing this out! I've tried it in TS playground and it doesn't pass this test.
From the top of my mind I was thinking about something like this test:
```ts
type IsConstructable<T> = T extends new (...args: any[]) => any ? true : false
```

But seem to work for `typeof MyClass`. I'll try and find a better solution, or we just keep it as you've suggested. See [here](https://www.typescriptlang.org/play?ts=4.9.5#code/C4TwDgpgBAkgzgYQPYDs7AE4FcDGwCGARgDYQA8AKgHxQC8UFUEAHsBCgCZxQoQDuUABQA6UfgwBzOAC4o+FCADaAXQCUdGvJBQA-FExZosgGb5icCAChLOYvjjcAqhYxQA3paheoxgJYZ0ADl8AFsIAEJZdAxfFAlPbzsg0IiozFj4gF9rHFR0KCwXOh5+KGcIDEFVa1BIKAARfDZ4ZDQDPCJSYpa89oIScka2Kkta6AQkuB627A6B7sRe2f7SMjGkYzKXEbHYNvkcCGno3BXoemO+zvJ1zcKKkaA).

Reply:
Not to mention some of the builtin objects technically constructable: like Date, Map and Set